### PR TITLE
Allowing direct CEL reserved keyword usage in CRD

### LINKED ISSUE #122777: CEL has misleading error message for fields called "namespace"
### What happened?

See the CRD at https://gist.github.com/mattcary/353b202240c250b5cb4371058f233476

in summary it is

```
spec:
  ref:
    name: myName
    namespace: myNamespace
```

One can create a x-kubernetes-validations rule that spec.ref.name != "", but one cannot create a rule that spec.ref.namespace != "". The former applies successfully, but applying the latter (the gist as written now):

```
The CustomResourceDefinition "validations.mattcary.info" is invalid: spec.validation.openAPIV3Schema.properties[spec].x-kubernetes-validations[0].rule: Invalid value: apiextensions.ValidationRule{Rule:"self.ref.namespace != \"\"", Message:"", MessageExpression:"", Reason:(*apiextensions.FieldValueErrorReason)(nil), FieldPath:"", OptionalOldSelf:(*bool)(nil)}: compilation failed: ERROR: <input>:1:9: undefined field 'namespace'
 | self.ref.namespace != ""
 | ........^
```

### What did you expect to happen?

The error message should say improper use of reserved word "namespace" rather than say that the field does not exist.

### How can we reproduce it (as minimally and precisely as possible)?

Save the gist https://gist.github.com/mattcary/353b202240c250b5cb4371058f233476 to a file, then `kubectl apply -f FILENAME`

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.0-gke.1492000
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux rodete"
NAME="Debian GNU/Linux rodete"
VERSION_CODENAME=rodete
ID=debian
HOME_URL="https://go/glinux"
SUPPORT_URL="https://go/techstop"
BUG_REPORT_URL="https://go/techstop"
$ uname -a
Linux mattcary.c.googlers.com 6.5.13-1rodete1-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.5.13-1rodete1 (2023-12-06) x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by ballista01**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR addresses a critical usability issue where CRD validation using CEL could not directly reference fields with names that are reserved keywords in CEL, such as namespace. By adding a check to handle reserved keywords without requiring them to be escaped, we enhance the developer experience by allowing direct reference to these fields in validation rules. This change aligns with the intuitive use of field names and removes the need for workarounds or confusing error messages related to field name usage.

#### Which issue(s) this PR fixes:

Fixes [#122777](https://github.com/kubernetes/kubernetes/issues/122777)

#### Special notes for your reviewer:

Following @thockin's advice, this solution does not introduce additional error messages for using reserved keywords as field names. Instead, it implements a check to support the direct use of reserved keywords, improving clarity and usability in CRD validation rules. The implementation has been tested to ensure backward compatibility and to avoid introducing breaking changes.

#### Does this PR introduce a user-facing change?

```release-note
CRD validation rules now support direct use of reserved keywords as field names in CEL expressions, enhancing usability and developer experience.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[CRD validation rule doc](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
-->
[CEL reserved keyword](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)
```

### CODE REVIEW

**Comment by ballista01** [comment]
Manual test seems fine:
image
@thockin @cici37 @mattcary Does this fix look good to you?

**Comment by mattcary** [comment]
Very cool! Thank you for going the extra step to allow reserved words.

And thanks to @thockin for pushing for the right fix.

/lgtm

**Comment by thockin** [comment]
Thanks!  This needs a careful look by a proper CEL expert, of course :)

@jpbetz  @cici37

**Comment by sftim** [comment]
Changelog suggestion
```diff
-CRD validation rules now support direct use of reserved keywords as field names in CEL expressions, enhancing usability and developer experience.
+Revised CRD validation rules. These rules now support direct use of CEL reserved identifiers (such as `function`
+or `namespace`) as field names within CEL expressions, enhancing usability and user experience.
```

(those identifiers were reserved to allow easier embedding of CEL into a host language, so it's fine to allow them as field names)

**Comment by jpbetz** [comment]
This is a change needs to be ratcheted in. We need to Allow persisted CEL expressions to contain keywords one release before we allow API requests to CEL expressions  that contain keywords.

File: staging/src/k8s.io/apiserver/pkg/cel/types.go [hunk 1]
```diff
@@ -452,6 +452,10 @@ func (rt *DeclTypeProvider) FindFieldType(typeName, fieldName string) (*ref.Fiel
 	}
 
 	f, found := st.Fields[fieldName]
+	if !found {
+		f, found = st.Fields["__"+fieldName+"__"]
+	}
+
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/cel/types.go:458]: Let's instead change the code that writes values to Fields to not escape values.  This approach is problematic in the pathological case (use has a field named `if` and a field named `__if__`).

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go [hunk 2]
```diff
@@ -8251,35 +8251,19 @@ func TestValidateCustomResourceDefinitionValidation(t *testing.T) {
...
 			input: apiextensions.CustomResourceValidation{
 				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
 					Type: "object",
 					XValidations: apiextensions.ValidationRules{
-						{
-							Rule: "self.if > 0",
-						},
-						{
-							Rule: "self.namespace > 0",
-						},
```
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go:8264]: Are ALL cel keywords safe?  If so we should verify them all with test..

**Comment by jpbetz** [comment]
I'd be interested in seeing if we can make progress on this. I've added a couple comments.  We would need to gate it so that we support it in the CEL environment for at lease one release before we allow usages of it.

cc @cici37 @aojea

**Comment by cici37** [comment]
Hi @ballista01 , Thanks for raising this PR! Would you still be willing to carry this forward? It would be great if we could have this change merged in 1.31 :)

File: staging/src/k8s.io/apiserver/pkg/cel/types.go (hunk 1 above)
Comment (cici37) [reply staging/src/k8s.io/apiserver/pkg/cel/types.go:458]: The problem is existing without the current PR: https://github.com/kubernetes/kubernetes/issues/126132 so this doesn't introduce new issue :)

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/cel/types.go:458]: Can we narrow this down to not-already-escaped names?  I think that's enough to take care of problems related to multiple-levels of escaping.  For example:
```go
if !found && !strings.HasPrefix(fieldName, "__") {
...
}
```

**Comment by jpbetz** [comment]
We need to ratchet this in, otherwise this breaks when a `namespace` field is written to a 1.31 cluster and then the cluster is rolled back.

^ @cici37

**Comment by liggitt** [comment]
/close
in favor of https://github.com/kubernetes/kubernetes/pull/126188

**Pull request closed without merging**